masterplan_tools/models/city_model.py

```diff
@@ -6,136 +6,228 @@
-    """in meters. Objects with smaller area will be dropped."""
-
-    def __init__(  # pylint: disable=too-many-arguments
+    blocks: PolygonGeoJSON[CityBlockFeature]
+    accessibility_matrix: AccessibilityMatrix
+    services: dict[str, PointGeoJSON[ServicesFeature]]
+    services_graph: InstanceOf[nx.Graph] = Field(exclude=True, default=None)
+
+    def get_service_types(self) -> list[str]:
+        return list(self.services.keys())
+
+    @field_validator("blocks", mode="before")
+    def validate_blocks(value):
+        if isinstance(value, gpd.GeoDataFrame):
+            return PolygonGeoJSON[CityBlockFeature].from_gdf(value)
+        return value
+
+    @field_validator("accessibility_matrix", mode="before")
+    def validate_matrix(value):
+        if isinstance(value, pd.DataFrame):
+            return AccessibilityMatrix(df=value)
+        return value
+
+    @field_validator("services", mode="before")
+    def validate_services(value):
+        dict = value.copy()
+        for service_type in dict:
+            if isinstance(dict[service_type], gpd.GeoDataFrame):
+                dict[service_type] = PointGeoJSON[ServicesFeature].from_gdf(value[service_type])
+        return dict
+
+    def visualize(self, max_distance=7) -> None:
+        blocks = self.blocks.to_gdf()
+        centroids = blocks.copy()
+        centroids["geometry"] = centroids["geometry"].centroid
+        edges = []
+        for u, v, a in self.services_graph.edges(data=True):
+            if a["weight"] < max_distance and a["weight"] > 0:
+                edges.append(
+                    {
+                        "distance": a["weight"],
+                        "geometry": LineString([centroids.loc[u, "geometry"], centroids.loc[v, "geometry"]]),
+                    }
+                )
+        edges = gpd.GeoDataFrame(edges).sort_values(ascending=False, by="distance")
+        fig, ax = plt.subplots(figsize=(15, 15))
+        blocks.plot(ax=ax, alpha=0.5, color="#ddd")
+        edges.plot(ax=ax, alpha=0.1, column="distance", cmap="summer")
+        plt.show()
+
+    def prepare_graph(  # pylint: disable=too-many-arguments,too-many-locals
         self,
-        buildings: gpd.GeoDataFrame | None = None,
-        services: dict[gpd.GeoDataFrame] = ...,
-        roads_geometry: gpd.GeoDataFrame | None = None,
-        water_geometry: gpd.GeoDataFrame | None = None,
-        railways_geometry: gpd.GeoDataFrame | None = None,
-        nature_geometry_boundaries: gpd.GeoDataFrame | None = None,
-        city_geometry: gpd.GeoDataFrame | None = None,
-        accessibility_matrix: gpd.GeoDataFrame = ...,
-        transport_graph: nx.Graph | None = None,
-        greenings: gpd.GeoDataFrame | None = None,
-        parkings: gpd.GeoDataFrame | None = None,
-        city_blocks: gpd.GeoDataFrame = ...,
-    ) -> None:
-        """Initialize CityModel
+        service_type: str,
+        services_graph: nx.Graph,
+        updated_block_info: dict = None,
```

Comment:
`None` is not a valid default for a `dict` type (which is better to be typed as well)